denyhosts.8
.TH DENYHOSTS "8" "February 2015" "DenyHosts version: 2.10" "User Commands"
.SH NAME
DenyHosts \- version: 2.10
.SH DESCRIPTION
.B DenyHosts
is a python program that automatically blocks ssh attacks by
adding entries to /etc/hosts.deny.
.B DenyHosts
will also inform system administrators about offending hosts,
attacked users and suspicious logins.

Usage:
/usr/sbin/denyhosts [\-f logfile | \fB\-\-file\fR=\fIlogfile]\fR [ \fB\-c\fR configfile | \fB\-\-config\fR=\fIconfigfile]\fR [\-i | \fB\-\-ignore]\fR [\-n | \fB\-\-noemail]\fR [\-\-purge] [\-\-migrate] [\-\-daemon] [\-\-sync] [\-\-version]
.TP
\fB\-\-file\fR:
The name of log file to parse
.HP
\fB\-\-config\fR: The configuration file to use, typically /etc/denyhosts.conf
.HP
\fB\-\-ignore\fR: Ignore last processed offset (start processing from beginning)
.HP
\fB\-\-noemail\fR: Do not send an email report
.HP
\fB\-\-unlock\fR: if lockfile exists, remove it and run as normal
.HP
\fB\-\-migrate\fR: migrate your HOSTS_DENY file so that it is suitable for \fB\-\-purge\fR
.HP
\fB\-\-purge\fR: expire entries older than your PURGE_DENY setting
.HP
\fB\-\-purge\-all\fR: expire all entries immediately
.HP
\fB\-\-purgeip\fR: remove specified IP address from blocked list
.HP
\fB\-\-daemon\fR: run DenyHosts in daemon mode
.HP
\fB\-\-foreground\fR: run DenyHosts in foreground mode
.HP
\fB\-\-sync\fR: run DenyHosts synchronization mode
.HP
\fB\-\-version\fR: Prints the version of DenyHosts and exits
.HP
\fB\-\-test-email\fR: Sends test email according to config and exits
.PP
Note: multiple \fB\-\-file\fR args can be processed.  If multiple files are provided, \fB\-\-ignore\fR is implied
.PP
Note: in a Debian system, the default running mode is \fBdaemon mode\fR and the configuration file is \fB/etc/denyhosts.conf\fR
.SS "When run in --daemon mode the following flags are ignored:"
.HP
\fB\-\-file\fR, \fB\-\-purge\fR, \fB\-\-migrate\fR, \fB\-\-sync\fR, \fB\-\-verbose\fR

.SH "SEE ALSO"
Please refer to http://denyhost.sourceforge.net/faq.html for full documentation.
.SH AUTHOR
.B DenyHosts
was written by Phil Schwartz <phil_schwartz@users.sourceforge.net>
.PP
This manual page was written by Marco Bertorello <marco@bertorello.ns0.it>
for the Debian project (but may be used by others).


CHANGELOG.txt
DENYHOSTS CHANGELOG

3.1
======================

Fixed a type check in DenyHosts/report.py which was causing
problems when moving between Python2 and Python3.

Added checks to see if an IP address is valid. This
pulls in the requirement for the ipaddr Python module.

Added check to see if there is a break-in attempt against
the Dovecot imap service. This is an option which can be
enabled/disabled in the configuration file. It is turned off
by default.



3.0
======================

Initial translation of code from Python 2 to Python 3.
DenyHosts can now be run as either a Python 2 or a Python 3
program.

Added patch from Fedora to fix initial sync issue and
insure info logging stream is active.
(Provided by Jason Tibbitts.)

Added "import logging" to denyhosts.py to avoid errors
when setting up logging. (See above change.)

Added option PF_TABLE_FILE to the configuration file.
When this option is enabled it causes DenyHosts to write
blocked IP addresses to a text file. The default location
is /etc/blacklist. This text file should correspond to a
PF firewall table.

At start-up, try to create the file specified by
HOSTS_DENY. That way we avoid errors later if the
file does not exists. Can be a problem on operating systems
where /etc/hosts.deny does not exist in the default
configuration.

Added regex pattern to detect invalid user accounts. This blocks
connections from remote hosts who are attempting to login
with accounts not found on the local system.
While these connections to non-existent accounts are relatively harmless, 
they are usually used as part of a brute force attack and filtering them
before they reach OpenSSH is a good idea.



2.10
======================

- Updated example rule for PF in configuration file
  to make black listing attacking IPs more effective.

- Added debugging info in case we cannot create a new
  PF table entry.

- Fixed syntax for comparing suspecious logins. Avoids
  always testing true/false depending on Python version.

- No longer require ETC_DIR in the configuration file.
  Use a default value "/etc" if ETC_DIR is not manually
  specified.

- Make sure DenyHosts logs when running in foreground mode.
  When in foreground, warnings are logged to a file rather
  than outputted to terminal. Keeps things clean.

- Add --unlock command line arguement to remove old
  lock files.

- Updated README, version and Makefile with new 
  version/maintainer information.

- Added check for PAM failures on FreeBSD. This should block both
  failed user logins that are reported by PAM and also block
  repteated attempts at accessing the root account when root
  logins are disabled by OpenSSH. The latter does not really add
  more practical protection, but can prevent the connection \
  attempts at the firewall level before the OpenSSH service
  is contacted.

- Add systemd unit file, denyhosts.service



2.9 (November 3, 2014)
======================

- DenyHost now supports working with the PF
  packet filter, a popular firewall for FreeBSD,
  OpenBSD, TrueOS, PC-BSD and NetBSD.
  To enable PF support in DenyHost, comment
  out the IPTABLES option in the denyhosts.conf file
  and enable the PFCTL_PATH and PF_TABLE options.

  DenyHost will add misbehaving IP addresses to the
  PF table specified by "PF_TABLE". This table
  should be blocked using the pf.conf file. Please
  see the denyhosts.conf file for more information
  and example PF rules for blocking incoming traffic.

  Please note that even if /etc/hosts.deny is not used
  to block incoming connectins, the file should still exists
  or DenyHosts may throw an error. (This should be fixed
  in the next release.)



2.8 (June 12, 2014)
===================

- Use standard errno instead of hardcoded errno value.
  Patch provided by Pino Toscano.

- Make sure PLUGIN_DENY is called for each host we receive from
  the sync server.
  Patch provided by Sean M. Collins.

- Made sure only new hosts in hosts.deny are reported as new, not
  all hosts. This prevents the PLUGIN_DENY plugin from getting
  old entries repeatedly.
  Patch provided by Chris Erdle.

- We now check user defined regular expression filters, even
  if we already found a match with an existing filter. This
  allows the user to filter more services without using
  a plugin.
  Patch provided by Ben.

- Added --purge-all command line flag to allow us to remove all
  old entries from the deny file without waiting.
  Patch provided by 9MediaCenterGUI on SourceForge.

- Updated copyright information and some documentation.

- Added manual page from Debian and fixed typo. Added
  additional command line options to man page.

- Added --purgeip option to allow us to remove specific
  IP addresses from the blocked list at start time.
  Patch provided by Nelson Howell.
  Should close Debian bug 529089.

- Updated FAILED_ENTRY_REGEX7 to be more flexible.

- Added ability to use Linux iptables to block incoming
  connections. See IPTABLES option in the configuration file.

- Made it possible to block specific ports, allowing remote
  hosts to conenct to some services while being blocked on
  others by the iptables firewall.
  See the BLOCKPORT option in the configuration file.



2.7 (May 18, 2014)
==================

- Forked code from DenyHosts (denyhosts.sf.net)
  New project now maintained at denyhost.sf.net

- Added private moduls patch from Marco Bertorello. Loads
  modules from /usr/share/denyhosts

- Place config, lock and executable file in more
  standard locations. Patch provided by Marco Bertorello.

- Fixed configuration (denyhosts.cfg-dist) to better support
  Debian and Ubuntu. Patch supplied by Marco Nenciarini.

- Added warning to migrate switch. Patch provided by
  Marco Bertorello.

- Avoid installing unwanted files (extra scripts and changelog).
  Patch provided by Marco Nenciarini.

- Fix bug which would not recognize an attack on the root
  user account. Patch provided by Kyle Willmon.

- Fix pattern matching bug (CVE-2007-4323).

denyhosts.py
#!/usr/bin/env python
import os
import platform
import sys
import logging
sys.path.insert(0, '/usr/share/denyhosts')

import DenyHosts.python_version

import getopt
from getopt import GetoptError
import traceback

from DenyHosts.util import die, setup_logging, is_true, send_email
from DenyHosts.lockfile import LockFile
from DenyHosts.prefs import Prefs
from DenyHosts.version import VERSION
from DenyHosts.deny_hosts import DenyHosts
from DenyHosts.denyfileutil import Purge, PurgeIP, Migrate, UpgradeTo099
from DenyHosts.constants import *
from DenyHosts.sync import Sync

info = logging.getLogger("denyhosts").info
#################################################################################

def usage():
    print("Usage:")
    print('{0} [-f logfile | --file=logfile] [ -c configfile | --config configfile] '
          '[-i | --ignore] [-n | --noemail] [--purge] [--purge-all] [--purgeip ip] '
          '[--migrate] [--daemon] [--sync] [--version]'.format(sys.argv[0]))
    print("\n\n")
    print(" --config: The pathname of the configuration file")
    print(" --file:   The name of log file to parse")
    print(" --ignore: Ignore last processed offset (start processing from beginning)")
    print(" --noemail: Do not send an email report")
    print(" --unlock: if lockfile exists, remove it and run as normal")
    print(" --migrate: migrate your HOSTS_DENY file so that it is suitable for --purge")
    print(" --purge: expire entries older than your PURGE_DENY setting")
    print(" --purge-all: expire all entries")
    print(" --purgeip: expire designated IP entry immediately")
    print(" --daemon: run DenyHosts in daemon mode")
    print(" --foreground: run DenyHosts in foreground mode")
    print(" --sync: run DenyHosts synchronization mode")
    print(" --version: Prints the version of DenyHosts and exits")
    print(" --test-email: Sends test email according to config and exits")

    print("\n")
    print("Note: multiple --file args can be processed. ")
    print("If multiple files are provided, --ignore is implied")
    print("\n")
    print("Note: multiple --purgeip arguments can be processed. ")
    print("\n")
    print("When run in --daemon mode the following flags are ignored:")
    print("     --file, --purge, --purge-all, --purgeip, --migrate, --sync, --verbose")


#################################################################################
#################################################################################


if __name__ == '__main__':
    logfiles = []
    purgeip_list = []
    config_file = CONFIG_FILE
    ignore_offset = 0
    noemail = 0
    verbose = 0
    migrate = 0
    purge = 0
    purge_all = 0
    sync_mode = 0
    daemon = 0
    foreground = 0
    enable_debug = 0
    purgeip = 0
    upgrade099 = 0
    unlock = 0
    args = sys.argv[1:]
    try:
        (opts, getopts) = getopt.getopt(args, 'f:c:dinuvps?hV',
                                        ["file=", "ignore", "verbose", "debug", "test-email",
                                         "help", "noemail", "config=", "version",
                                         "migrate", "purge", "purge-all", "purgeip", "daemon", "foreground",
                                         "unlock", "sync", "upgrade099"])
    except GetoptError:
        print("\nInvalid command line option detected.")
        usage()
        sys.exit(1)

    for opt, arg in opts:
        if opt in ('-h', '-?', '--help'):
            usage()
            sys.exit(0)
        if opt in ('-f', '--file'):
            logfiles.append(arg)
        if opt in ('-i', '--ignore'):
            ignore_offset = 1
        if opt in ('-n', '--noemail'):
            noemail = 1
        if opt in ('-v', '--verbose'):
            verbose = 1
        if opt in ('-d', '--debug'):
            enable_debug = 1
        if opt in ('-c', '--config'):
            config_file = arg
        if opt in ('-m', '--migrate'):
            migrate = 1
        if opt in ('-p', '--purge'):
            purge = 1
        if opt in ('-s', '--sync'):
            sync_mode = 1
        if opt in ('-s', '--unlock'):
            unlock = 1
        if opt == '--daemon':
            daemon = 1
        if opt == '--foreground':
            foreground = 1
        if opt == '--purge-all':
            purge_all = 1
        if opt == '--purgeip':
            purgeip_list.append(arg)
            purgeip = 1
        if opt == '--upgrade099':
            upgrade099 = 1
        if opt == '--version':
            print("DenyHosts version:", VERSION)
            sys.exit(0)
        if opt == '--test-email':
            send_email(Prefs(config_file), "Testing DenyHosts email")
            sys.exit(0)

    # This is generally expected to be in the environment, but there's no
    # non-hackish way to get systemd to set it, so just hack it in here.
    os.environ['HOSTNAME'] = platform.node()

    prefs = Prefs(config_file)

    first_time = 0
    try:
        if not os.path.exists(prefs.get('WORK_DIR')):
            os.makedirs(prefs.get('WORK_DIR'))
            first_time = 1
    except Exception as e:
        if e[0] != 17:
            print(e)
            sys.exit(1)

    # On some operating systems the file /etc/hosts.deny (or
    # whatever HOSTS_DENY is set to, may not exist. We should
    # "touch" the file to make sure it is there to avoid errors later.
    try:
        host_filename = prefs.get("HOSTS_DENY")
        if host_filename:
            fp = open(prefs.get("HOSTS_DENY"), "a")
            fp.close()
    except Exception as e:
        print("Unable to create file specified by HOSTS_DENY variable.")

    setup_logging(prefs, enable_debug, verbose, daemon)

    if not logfiles or daemon:
        logfiles = [prefs.get('SECURE_LOG')]
    elif len(logfiles) > 1:
        ignore_offset = 1

    if not prefs.get('ADMIN_EMAIL'):
        noemail = 1

    lock_file = LockFile(prefs.get('LOCK_FILE'))

    if unlock:
        if os.path.isfile(prefs.get('LOCK_FILE')):
            lock_file.remove()

    lock_file.create()

    if upgrade099 and not (daemon or foreground):
        if not prefs.get('PURGE_DENY'):
            lock_file.remove()
            die(
                "You have supplied the --upgrade099 flag," +
                " however you have not set PURGE_DENY in your configuration file"
            )
        else:
            u = UpgradeTo099(prefs.get("HOSTS_DENY"))

    if migrate and not (daemon or foreground):
        if not prefs.get('PURGE_DENY'):
            lock_file.remove()
            die("You have supplied the --migrate flag however you have not set PURGE_DENY in your configuration file.")
        else:
            m = Migrate(prefs.get("HOSTS_DENY"))

    # clear out specific IP addresses
    if purgeip and not daemon:
        if len(purgeip_list) < 1:
            lock_file.remove()
            die("You have provided the --purgeip flag however you have not listed any IP addresses to purge.")
        else:
            try:
                p = PurgeIP(
                    prefs,
                    purgeip_list
                )

            except Exception as e:
                lock_file.remove()
                die(str(e))

    # Try to purge old records without any delay
    if purge_all and not daemon:
        purge_time = 1
        try:
            p = Purge(prefs, purge_time)
        except Exception as e:
            lock_file.remove()
            die(str(e))

    if purge and not (daemon or foreground):
        purge_time = prefs.get('PURGE_DENY')
        if not purge_time:
            lock_file.remove()
            die("You have provided the --purge flag however you have not set PURGE_DENY in your configuration file.")
        else:
            try:
                p = Purge(prefs,
                          purge_time)

            except Exception as e:
                lock_file.remove()
                die(str(e))

    try:
        for f in logfiles:
            dh = DenyHosts(f, prefs, lock_file, ignore_offset,
                           first_time, noemail, daemon, foreground)
    except KeyboardInterrupt:
        pass
    except SystemExit as e:
        pass
    except Exception as e:
        traceback.print_exc(file=sys.stdout)
        print("\nDenyHosts exited abnormally")

    if sync_mode and not (daemon or foreground):
        if not prefs.get('SYNC_SERVER'):
            lock_file.remove()
            die("You have provided the --sync flag however your configuration file is missing a value for SYNC_SERVER.")
        sync_upload = is_true(prefs.get("SYNC_UPLOAD"))
        sync_download = is_true(prefs.get("SYNC_DOWNLOAD"))
        if not sync_upload and not sync_download:
            lock_file.remove()
            die(
                "You have provided the --sync flag"
                " however your configuration file has SYNC_UPLOAD and SYNC_DOWNLOAD set to false."
            )
        try:
            sync = Sync(prefs)
            if sync_upload:
                timestamp = sync.send_new_hosts()
            if sync_download:
                new_hosts = sync.receive_new_hosts()
                if new_hosts:
                    # MMR: What is 'info' here?
                    info("received new hosts: %s", str(new_hosts))
                    # TODO sync method's don't exist what should these be doing?
                    sync.get_denied_hosts()
                    sync.update_hosts_deny(new_hosts)

                    dh.get_denied_hosts()
                    dh.update_hosts_deny(new_hosts)
            sync.xmlrpc_disconnect()
        except Exception as e:
            lock_file.remove()
            die("Error synchronizing data", e)

    # remove lock file on exit
    lock_file.remove()
